File: VB/E2371.Module/CustomAuthenticationStandardProvider.vb
Imports Microsoft.VisualBasic
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Security
Imports E2371.Module.BusinessObjects
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace E2371.Module
	Public Class CustomAuthenticationStandardProvider
		Inherits AuthenticationStandardProvider
		Public Sub New(ByVal userType As Type)
			MyBase.New(userType)
		End Sub
		Public Overrides Function Authenticate(ByVal objectSpace As IObjectSpace) As Object
			Dim customPermissionPolicyUser As CustomPermissionPolicyUser = Nothing
			Try
				customPermissionPolicyUser = TryCast(MyBase.Authenticate(objectSpace), CustomPermissionPolicyUser)
			Catch e As AuthenticationException
				ThrowAuthenticationError(e.UserName)
			End Try
			If customPermissionPolicyUser IsNot Nothing AndAlso customPermissionPolicyUser.IsActiveDirectoryUser Then
				ThrowAuthenticationError(customPermissionPolicyUser.UserName)
			End If
			Return customPermissionPolicyUser
		End Function

		Private Shared Sub ThrowAuthenticationError(ByVal userName As String)
			Throw New Exception(String.Format("Login failed for '{0}'. Make sure your user name is correct and retype the password in the correct case, or use the Windows authentication instead.", userName))
		End Sub
	End Class
End Namespace

File: CS/E2371.Module/CustomAuthenticationStandardProvider.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Security;
using E2371.Module.BusinessObjects;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace E2371.Module {
    public class CustomAuthenticationStandardProvider : AuthenticationStandardProvider {
        public CustomAuthenticationStandardProvider(Type userType) : base(userType) {
        }
        public override object Authenticate(IObjectSpace objectSpace) {
            CustomPermissionPolicyUser customPermissionPolicyUser = null;
            try {
                customPermissionPolicyUser = base.Authenticate(objectSpace) as CustomPermissionPolicyUser;
            }
            catch(AuthenticationException e) {
                ThrowAuthenticationError(e.UserName);
            }
            if(customPermissionPolicyUser != null && customPermissionPolicyUser.IsActiveDirectoryUser) {
                ThrowAuthenticationError(customPermissionPolicyUser.UserName);
            }
            return customPermissionPolicyUser;
        }

        private static void ThrowAuthenticationError(string userName) {
            throw new Exception(string.Format("Login failed for '{0}'. Make sure your user name is correct and retype the password in the correct case, or use the Windows authentication instead.", userName));
        }
    }
}
